agent/consul/acl_replication.go

```diff
@@ -1,347 +1,556 @@
+			return false, fmt.Errorf("Failed to apply policy deletions: %v", respErr)
+		}
+
+		if i+aclBatchDeleteSize < len(deletions) {
+			select {
+			case <-stopCh:
+				return true, nil
+			case <-ticker.C:
+				// do nothing - ready for the next batch
+			}
+		}
+	}
+
+	return false, nil
 }
 
-// See sort.Interface.
-func (a *aclIterator) Swap(i, j int) {
-	a.acls[i], a.acls[j] = a.acls[j], a.acls[i]
+func (s *Server) updateLocalACLPolicies(policies structs.ACLPolicies, stopCh <-chan struct{}) (bool, error) {
+	ticker := time.NewTicker(time.Second / time.Duration(s.config.ACLReplicationApplyLimit))
+	defer ticker.Stop()
+
+	// outer loop handles submitting a batch
+	for batchStart := 0; batchStart < len(policies); {
+		// inner loop finds the last element to include in this batch.
+		batchSize := 0
+		batchEnd := batchStart
+		for ; batchEnd < len(policies) && batchSize < aclBatchUpsertSize; batchEnd += 1 {
+			batchSize += policies[batchEnd].EstimateSize()
+		}
+
+		req := structs.ACLPolicyBatchUpsertRequest{
+			Policies: policies[batchStart:batchEnd],
+		}
+
+		resp, err := s.raftApply(structs.ACLPolicyUpsertRequestType, &req)
+		if err != nil {
+			return false, fmt.Errorf("Failed to apply policy upserts: %v", err)
+		}
+		if respErr, ok := resp.(error); ok && err != nil {
+			return false, fmt.Errorf("Failed to apply policy upsert: %v", respErr)
+		}
+		s.logger.Printf("[DEBUG] acl: policy replication - upserted 1 batch with %d policies of size %d", batchEnd-batchStart, batchSize)
+
+		// policies[batchEnd] wasn't include as the slicing doesn't include the element at the stop index
+		batchStart = batchEnd
+
+		// prevent waiting if we are done
+		if batchEnd < len(policies) {
+			select {
+			case <-stopCh:
+				return true, nil
+			case <-ticker.C:
+				// nothing to do - just rate limiting
+			}
+		}
+	}
+	return false, nil
 }
 
-// See sort.Interface.
-func (a *aclIterator) Less(i, j int) bool {
-	return a.acls[i].ID < a.acls[j].ID
+func (s *Server) fetchACLPoliciesBatch(policyIDs []string) (*structs.ACLPoliciesResponse, error) {
+	req := structs.ACLPolicyBatchReadRequest{
+		Datacenter: s.config.ACLDatacenter,
+		PolicyIDs:  policyIDs,
+		QueryOptions: structs.QueryOptions{
+			AllowStale: true,
+			Token:      s.tokens.ACLReplicationToken(),
+		},
+	}
+
+	var response structs.ACLPoliciesResponse
+	if err := s.RPC("ACL.PolicyBatchRead", &req, &response); err != nil {
+		return nil, err
+	}
+
+	return &response, nil
 }
 
-// Front returns the item at index position, or nil if the list is exhausted.
-func (a *aclIterator) Front() *structs.ACL {
-	if a.index < len(a.acls) {
-		return a.acls[a.index]
+func (s *Server) fetchACLPolicies(lastRemoteIndex uint64) (*structs.ACLPolicyListResponse, error) {
+	defer metrics.MeasureSince([]string{"leader", "replication", "acl", "policy", "fetch"}, time.Now())
+
+	req := structs.ACLPolicyListRequest{
+		Datacenter: s.config.ACLDatacenter,
+		QueryOptions: structs.QueryOptions{
+			AllowStale:    true,
```

Comment:
@kyhavlov this also reminds me - did you allow replication using stale?

I'm in two minds about this @mkeeler, at first it seemed like an obvious win since we are doing async replication anyway but actually the semantics of our stale are more subtle.

For example in the case that the primary DC has 5 servers and one of them is partitioned from the others but not from the WAN and might be hours or days stale, this code could blindly just keep blocking on that server and the whole secondary DC would be behind until it rebalances to another server or the partition heals.

That's _probably_ OK, but perhaps we should use stale = true but then actually verify the last contact time was in some small bound and re-issue requests to leader if not? Actually the ideal would be not to re-issue to leader but to re-issue to a random other server so load remains spread and we don't suddenly send all the load to the leader if the primary cluster gets unhealthy which could exacerbate an outage. 🤔 I *think* that is possible by calling `RebalanceServers` and then retrying the RPC with `AllowState = true` in the case that stale result has a lastContact time that's too big (e.g. more than a few minutes). But then we'd need to be careful that if _all_ the servers are stale by more than the threshold, we should use their stale view anyway?

I think this is OK to punt a little if it's too complicated but we should certainly make an issue for this as I think it's important to get right for all these new replication loops.

I mentioned on Kyle's intention replication PR that it would be nice to not keep re-inventing the wheel with blocking query loops because we tend to miss a few of the subtle edge cases every time we do.

Reply:
Actually, we rebalance servers frequently enough that maybe that's OK as an upper bound. 2-3 mins (jittered) is the minimum time between rebalances and it only increases once cluster gets above about 10k clients _per server_ (i.e. ~30k per 3 server cluster and 50k for 5 servers). So in all practical clusters most of the time (excluding wild partitions although even then they need to be super wild) it's less than the 10 min blocking query timeout anyway.

Reply:
I will create a follow on item for this. Previous acl replication and nomads both use stale queries but I do think it would be a good idea to ensure its within some reasonable bound.